src/rtc.rs

```diff
@@ -3,131 +3,171 @@
+    }
 }
 
 impl Rtc {
-    pub fn rtc(rtc: RTC, apb1r1: &mut APB1R1, bdcr: &mut BDCR, pwrcr1: &mut pwr::CR1, clocks: Clocks) -> Self {
+    pub fn rtc(rtc: RTC, apb1r1: &mut APB1R1, bdcr: &mut BDCR, pwrcr1: &mut pwr::CR1, clocks: Clocks, rtc_config : RtcConfig) -> Self {
 
-        assert_eq!(clocks.lsi(), true); // make sure LSI is enabled
+        // assert_eq!(clocks.lsi(), true); // make sure LSI is enabled
         // enable peripheral clock for communication
         apb1r1.enr().modify(|_, w| w.rtcapben().set_bit());
         pwrcr1.reg().read(); // read to allow the pwr clock to enable
+
+        let mut rtc_struct = Self {
+            rtc,
+            rtc_config
+        };
+        rtc_struct.set_config(apb1r1, bdcr, pwrcr1, clocks, rtc_config);
+        
+        rtc_struct
         
-        pwrcr1.reg().modify(|_, w| w.dbp().set_bit());
-        while pwrcr1.reg().read().dbp().bit_is_clear() {}
         
-        bdcr.enr().modify(|_, w| { w.bdrst().set_bit() }); // reset
         
-        bdcr.enr().modify(|_, w| unsafe {
-            w.rtcsel()
-                /* 
-                    00: No clock
-                    01: LSE oscillator clock used as RTC clock
-                    10: LSI oscillator clock used as RTC clock
-                    11: HSE oscillator clock divided by 32 used as RTC clock 
-                */
-                .bits(0b10)
-                .rtcen()
-                .set_bit()
-                .bdrst() // reset required for clock source change
-                .clear_bit()
-        });
-
-
-       write_protection(&rtc, false);
-        {
-            init_mode(&rtc, true);
-            {
-                rtc.cr.modify(|_, w| unsafe {
-                    w.fmt()
-                        .clear_bit() // 24hr
-                        .osel()
-                        /* 
-                            00: Output disabled
-                            01: Alarm A output enabled
-                            10: Alarm B output enabled
-                            11: Wakeup output enabled 
-                        */
-                        .bits(0b00)
-                        .pol()
-                        .clear_bit() // pol high
-                });
-                
-                rtc.prer.modify(|_, w| unsafe {
-                    w.prediv_s()
-                        .bits(255)
-                        .prediv_a()
-                        .bits(127)
-                });
-            }
-            init_mode(&rtc, false);
-
-            // TODO configuration for output pins
-            rtc.or.modify(|_, w| {
-                w.rtc_alarm_type()
-                    .clear_bit()
-                    .rtc_out_rmp()
-                    .clear_bit()
-            });
-            
-        }
-        write_protection(&rtc, true);
-
-        Self {
-            rtc: rtc
-        }
     }
 
-    pub fn set_time(&self, time: &Time){
+    pub fn set_time(&self, time: Option<&Time>, date: Option<&Date>){
```

Comment:
Is there a reason for making this set both date and time?

I get why the `get_date_time()` has to return both, due to the concerns outlined in the PR description, but i am not a fan of this `set_time` API. 
I would either 
- leave it split as `set_date(&self, date: &Date)` and `set_time(&self, time: &Time)`
OR
- make it an non-optional: `set_date_time(&self, date: &Date, time: &Time)`

Reply:
It should most definitely be named set_date_time. I have actually done this to protect against the same case as with get_time. Since if you are setting the time to 23:59 and the sub seconds ticks over before you set the date, the date would be wrong.
Maybe making it none optional is the solution.